salt/modules/supervisord.py

```diff
@@ -317,7 +310,13 @@ def status_raw(name=None, user=None, conf_file=None, bin_env=None):
     ret = __salt__["cmd.run_all"](
         _ctl_cmd("status", name, conf_file, bin_env), runas=user, python_shell=False,
     )
-    return _get_return(ret)
+    retmsg = _get_return(ret)
+    # Supervisor>=4.0 sets non zero return code if all processes are not
+    # in RUNNING state. In this case, _get_return appends and "ERROR: " to
+    # the return, which we need to filter out
+    if retmsg.startswith("ERROR: "):
```

Comment:
What happens if there is a legitimate failure? Will the split occur correctly or fail? Will the user still get a useful error message? The other functions that use `_get_return` would need a solution as well. Can we parse the new return codes instead and only add ERROR for legitimate failures?

Reply:
@Ch3LL Thanks for looking at this. The main problem that we have is [this issue in supervisor](https://github.com/Supervisor/supervisor/issues/1223) where the `status` command has been modified to send non-zero return codes if queried process is not `RUNNING` or `STARTING`.

In case of legitimate failure too, the same workflow is being followed in the current releases - add `ERROR: ` to the return (if not already there) and try to parse it. The state code is responsible for making use of parsed output which displays the correct comment and fails the state. That behaviour will not change with this PR because the output passed by `module.salt.status` will still be preserved (minus the beginning `ERROR: ` which is anyways not checked for `supervisord.status` (It is checked for other commands which this PR does not touch)

Other functions which use `get_return` do not need to change because `start`, `stop` or other command's have not had their error code meanings changed in Supervisor-4.0 and because `state.supervisord` looks for ERROR in outputs of those functions.

As for parsing return codes, there is another open bug #50566 (in addtion to the one this PR fixes #58526). However, there is no direction from core as to how to solve this issue. However, I will summarize the options in hope that core will take notice & chose the best way forward to:

- supervisor maintainer has explicilty said the return code for `status` will not be rolled back so we can expect the change to be permanent. They recommend (I think) to whitelist return codes `0,3` for status (but there is no guarantee if they won't change whitelist in future and `3` also communicates `FATAL`).
- They also recommend textual parsing which will give correct status of each process. Salt already does this so return codes appear to be a bit superfluous and may be ignored in such a case.

Minus a long term fix, this is a quick fix which solves a very specific bug (if first process is STOPEED it cannot be restarted using salt state) which is an offshoot of the main issue (that `supervisorctl status` sets non-zero codes even when command succeeds).

Reply:
Thanks for the additional information. It helped add context around the issue. If 3 always means a process is not running,starting and is NOT an error why not just check for 0 AND 3 in `_get_return` instead?

Reply:
@Ch3LL That would indeed be the right thing to do if we are sure supervisor folks won't change the status codes further. From the discussion on the linked issue, I gather that they intend for `supervisorctl status` to convey status using return codes. So when we run `supervisorctl status XYZ`, we can parse status code & get its status. This implies status codes _may_ change in the future necessitating further tinkering in the `supervisord` module to adjust to new status codes. We could change to status code based parsing, but it is a mess when multiple process status are involved (`supervisorctl status all` or `supervisorctl status group:`), so salt's current strategy of doing text based parsing seems wise.

So it again brings us back to the two issues I mentioned in the previous comment (under options) - do we change code to allow some status codes or do we disregard status codes altogether?

I am inclined to fix the issue either way (although my preference is second option). I can close this PR (& linked issue as duplicate) & work on fixing the root cause depending on how salt-core wishes to resolve this.

FWIW, I think this discussion would be better suited on either of the issues (#50566 or #54028 after closing one as duplicate of other) so that consensus can be built & diverse opinions gathered.